lib/mastodon/cli_helper.rb

```diff
@@ -7,3 +7,48 @@
 ActiveJob::Base.logger       = dev_null
 HttpLog.configuration.logger = dev_null
 Paperclip.options[:log]      = false
+
+module Mastodon
+  module CLIHelper
+    def create_progress_bar(total = nil)
+      ProgressBar.create(total: total, format: '%c/%u |%b%i| %e')
+    end
+
+    def parallelize_with_progress(scope)
+      ActiveRecord::Base.configurations[Rails.env]['pool'] = options[:concurrency]
+
+      progress  = create_progress_bar(scope.count)
+      pool      = Concurrent::FixedThreadPool.new(options[:concurrency])
+      futures   = []
+      aggregate = Concurrent::AtomicFixnum.new(0)
+
+      scope.reorder(nil).find_each do |item|
+        progress.total = futures.size + 1 if progress.total < futures.size + 1
+
+        futures << Concurrent::Future.execute(executor: pool) do
```

Comment:
From what I understand, even though `find_each` work in batches, this code will go through *all* the items in the query and put them in the queue, which might not be what we want, wrt. RAM usage.

Would it be possible to reap finished jobs as soon as possible and block whenever the pool is busy?

Reply:
The futures start processing as soon as they are created, the map over values at the end is just to block execution until they all finish. What is stored are the futures and the return values, which are numbers in our case, not the ActiveRecord records, so it should help with memory usage.

Reply:
I'm still worried about it being too much if the tasks are slow. I guess it could be interesting to see how it runs for pruning old remote media on m.s